feat(client): migrate to json-rpc

KwilLuke opened the pull request:
The client.ts now uses kwil-db's JSONRPC endpoints (instead of REST).

BREAKING CHANGE: Kwil-JS now relies on kwil-db's JSONRPC endpoints (available in kwil-db v0.8+). From this commit on, you should only use Kwil-JS with kwil-db v0.8+.

KwilLuke committed:
feat(client): migrate to json-rpc

The client.ts now uses kwil-db's JSONRPC endpoints (instead of REST).

BREAKING CHANGE: Kwil-JS now relies on kwil-db's JSONRPC endpoints (available in kwil-db v0.8+).
From this commit on, you should only use Kwil-JS with kwil-db v0.8+.

KwilLuke commented on src/api_client/client.ts:
```diff
@@ -209,135 +188,154 @@ export default class Client extends Api {
-    throw new Error(
-      JSON.stringify(res.data) ||
+
+  switch (res.status) {
+    case 200:
+      break;
+    case 401:
+      throw new Error(JSON.stringify(res.data) || 'Unauthorized.');
+    case 404:
+      throw new Error(JSON.stringify(res.data) || 'Not found.');
+    case 500:
+      throw new Error(JSON.stringify(res.data) || 'Internal server error.');
+    default:
+      throw new Error(
+        JSON.stringify(res.data) ||
         'An unknown error has occurred.  Please check your network connection.'
-    );
+      );
+  }
+
+  if (!res.data) {
+    throw new Error(`failed to parse response: ${res}`);
+  }
+
+  if (res.data.error) {
+    const data = res.data.error.data ? `, data: ${JSON.stringify(res.data.error.data)}` : '';
+    throw new Error(`JSON RPC call error: code: ${res.data.error.code}, message: ${res.data.error.message}` + data);
+  }
+
+  if (res.data.jsonrpc !== '2.0') {
+    throw new Error(JSON.stringify(res.data) || 'Invalid JSON RPC response.');
   }
 
-  if (!selector) {
-    return { status: res.status, data: undefined };
+  if (!res.data.result) {
+    throw new Error(JSON.stringify(res.data) || 'No result in JSON RPC response.');
   }
 
   return {
```
@jchappelow, do you mind taking a look at this `checkRes` function? I tried to model similar to how you showed me [here](https://github.com/jchappelow/kwil-db/blob/9528cd89e2702be1bbcebd848d60ceaa95e8edb4/core/rpc/client/user/jsonrpc/client.go#L134-L177).

jchappelow commented on src/builders/payload_builder.ts:
```diff
@@ -371,19 +370,20 @@ Kwil Chain ID: ${tx.body.chain_id}
-        fee: newTx.body.fee?.toString(),
+        type: newTx.body.type as PayloadType,
+        // @ts-ignore - DO NOT MERGE WITHOUT RESOLVING - Jon is looking into why fee is unmarshalled to an integer.
+        fee: Number(newTx.body.fee),
```
Pretty sure the resolution is simple, but need to make sure this is as expected: https://github.com/kwilteam/kwil-db/pull/713

KwilLuke committed:
ci(ci.yaml): use json rpc

change CI to use kwil-db json RPC endpoints

KwilLuke committed:
feat(client): jsonrpc for KGW

Changed KGW APIs to use JSON RPC.

KwilLuke commented on src/core/auth.ts:
```diff
@@ -49,15 +49,15 @@ export function composeAuthMsg(
   }
   msg += '\n';
-  msg += `URI: ${target.href}\n`;
+  // @Yaiba: Should I trust the URI provided by KGW or should I use the domain / create my own?
```
@Yaiba, can you advise on this? Should I trust the URI provided by KGW or should I create my own?

Yaiba replied:
Since we have checked that the domain matches, KGW will always return `domain + /rpc/v1`, we can trust returned URI.

KwilLuke committed:
fix(client): individual logout on multi-user session

Fixed a minor bug that cleared all cookies on an individual logout for a multi-user session in
NodeKwil. Now, kwil-js appropriately manages all cookies for different sessions.

KwilLuke committed:
refactor: added `EncodableDatabase` type

Added `EncodableDatabase` type, which is to be used before RLP-encoding a schema. It allows
`rlp:nil` fields to be an empty array.

Yaiba commented:
Should we merge this? I have this branch tested kwilteam/kwil-db#746  in CI https://github.com/kwilteam/kwil-db/actions/runs/9161574685/job/25186730490 and passed

jchappelow commented:
> Should we merge this? I have this branch tested [kwilteam/kwil-db#746](https://github.com/kwilteam/kwil-db/pull/746) in CI https://github.com/kwilteam/kwil-db/actions/runs/9161574685/job/25186730490 and passed

Just a sec, scanning through the main handler

jchappelow commented:
Yeah, lgtm.  Fire when ready @KwilLuke
